### ISSUE TITLE
paddle.v2.layer.cross_entropy_cost和classification_cost的区别

### ISSUE THREAD

**Issue by cszhou** [PaddlePaddle/Paddle#3366]
两者的区别是？
我没查到`classification_cost`的具体的文档说明，所以这个的计算公式是什么？
`cross_entropy_cost`我看了文档，但是只看到文字描述，最终的计算公式是什么，给出的是一个batch的平均loss还是整体的？

**Comment by lcy-seso** [comment]
您好。
1. 在 Paddle 中 `classification_cost` 就是 `multiclass_cross_entropy`，可以两者之前并没有本质的区别，是一回事儿。
    - 他们的区别是：`classification_cost` 可以当做是对 `cross_entropy_cost` 的一个封装，它的内部定义了一个 `cross_entropy_cost` 以及一个用来评估错误率的evaluator
    - 代码在这里：https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer_config_helpers/layers.py#L4115
2. multiclass-cross-entropy 的计算公式是：对第$i$个样本，真实类别标签是 $y_i$ $\mathcal{L_i}=\log p(y_i)$，其中 $p(y)$ 是softmax层预测出来的属于各类的概率。
3. Paddle 中cost中计算的是当前batch上cost的平均值。

**Comment by cszhou** [comment]
@lcy-seso  Hi，我又来了。
我从paddle/v2/layer.py翻了下代码没看到，原来是放在paddle/trainer_config_helpers/layers.py。

你写的`multiclass_cross_entropy`的`loss`，应该是少了个负号？？

那么接下来问题又来了。

## 目前我实现的模型
输入`feature1`，`feature2，label`，其中`label`取值为`{0,1}`
`feature1`经过`embedding_1`层得到32维度的`emb_1`特征，
`feature2`经过`embedding_2`层得到32维度的`emb_2`特征，
`emb_1`和`emb_2`特征,经过`cos_sim`层得到`cos_fea`特征。
现阶段我使用的`mse_cost`，输入是`cos_fea`和`label`，由于`cos_fea`取值是[-1,1]，所以我将`label`映射到`{-1,1}`。这个已经完成了。

## 接下来打算实现的
接下来我打算使用`multiclass_cross_entropy`作为loss函数，问题出现了。`cos_fea`的输出是一个值，但是取值是[-1,1]，而不是[0,1]的概率。理想状况是，对`cos_fea`做一个变化 （cos_fea+1）/2的变化可以映射到[0,1]之间。但是由于`label`为{0,1}，所以`multiclass_cross_entropy`要求的input应该是两个概率输入。于是卡住了。
所以paddle有提供这种场景的解决方案而我不知道的么。如果没有，有什么建议？

**Comment by lcy-seso** [comment]
抱歉。。写的时候，确实少了一个符号。

**Comment by lcy-seso** [comment]
- multiclass_cross_entropy 必须和 softmax 激活配合使用。
- cos_sim 的输出是一个scalar，不能给cross entropy，想接cross entropy 特征输出至少应该大于等于2，

**Comment by cszhou** [comment]
@lcy-seso 

multiclass_cross_entropy 必须和 softmax 激活配合使用，这个我可以理解。但是softmax激活的意义就是给一个label对应的概率分布。

实际上，`cos_sim`的输出表示这两个特征相似与否，相似度高的为1，相似度低的为-1，我要预估的是点击率。所以`p1 = (cos_sim+1)/2`完全可以表示label为1的概率了啊。那么 `p2 = 1-p1`就是label为0的概率了。也就是说有了`p1`和`p2`我就有了softmax层输出的概率。

如果是用Caffe，我可以自定义一个loss层，实现`forward`功能和`backward`功能。paddle应该提供类似的自定义loss层的方案，具体是如何实现的呢？在哪里添加新层，哪里实现`forward`功能和`backward`功能。或者不用这么麻烦的方案？

**Comment by lcy-seso** [comment]
从你的描述，你需要的是类似Boosting算法里的指数损失函数，并不是 cross_entropy ，paddle 现在没有这个损失函数。cross_entropy 之前必须是softmax，概率有归一化要求。cos_sim 的计算结果，严格意义上不是概率。

**Comment by yuanfang-ab** [comment]
请问，multiclass_cross_entropy的负号有加上吗？有其他替代可以用吗？

**Comment by yuanxiangxie** [comment]
@cszhou Hi 你好，我有个需求也是需要对inference的结果做偏移压缩变换，但是现在遇到的问题是做完压缩偏移后，模型训练的时候event.metrics没有信息输出，请问你有遇到这个问题吗？
